client/src/query.rs

```diff
@@ -0,0 +1,302 @@
+                                        Response is neither a `ValidationFail` encoded value nor a valid utf-8 string error response. \
+                                        You are likely using a version of the client library that is incompatible with the version of the peer software",
+                                    ),
+                                Into::into
+                            ),
+                    )
+                },
+                ClientQueryError::Validation,
+            )),
+        _ => Err(ResponseReport::with_msg("Unexpected query response", resp).unwrap_or_else(core::convert::identity).into()),
+    }
+}
+
+fn decode_singular_query_response(
+    resp: &http::Response<Vec<u8>>,
+) -> QueryResult<SingularQueryOutputBox> {
+    let QueryResponse::Singular(resp) = decode_query_response(resp)? else {
+        return Err(eyre!(
+            "Got unexpected type of query response from the node (expected singular)"
+        )
+        .into());
+    };
+    Ok(resp)
+}
+
+fn decode_iterable_query_response(
+    resp: &http::Response<Vec<u8>>,
+) -> QueryResult<IterableQueryOutput> {
+    let QueryResponse::Iterable(resp) = decode_query_response(resp)? else {
+        return Err(eyre!(
+            "Got unexpected type of query response from the node (expected iterable)"
+        )
+        .into());
+    };
+    Ok(resp)
+}
+
+/// An iterable query cursor for use in the client
+#[derive(Debug)]
+pub struct ClientQueryCursor {
+    // instead of storing iroha client itself, we store the base URL and headers required to make a request
+    //   along with the account id and key pair to sign the request.
+    // this removes the need to either keep a reference or use an Arc, but breaks abstraction a little
+    request_head: ClientQueryRequestHead,
+    cursor: ForwardCursor,
+}
+
+/// Different errors as a result of query response handling
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum ClientQueryError {
+    /// Query validation error
+    Validation(#[from] ValidationFail),
+    /// Failed to constrain the query to a single result
+    Single(#[from] SingleQueryError),
+    /// Other error
+    Other(#[from] eyre::Error),
+}
+
+impl From<ResponseReport> for ClientQueryError {
+    #[inline]
+    fn from(ResponseReport(err): ResponseReport) -> Self {
+        Self::Other(err)
+    }
+}
+
+impl QueryExecutor for Client {
+    type Cursor = ClientQueryCursor;
+    type Error = ClientQueryError;
+    type SingleError = ClientQueryError;
+
+    fn execute_singular_query(
+        &self,
+        query: SingularQueryBox,
+    ) -> Result<SingularQueryOutputBox, Self::Error> {
+        let request_head = self.get_query_request_head();
+
+        let request = QueryRequest::Singular(query);
+
+        let response = request_head.assemble(request).build()?.send()?;
+        let response = decode_singular_query_response(&response)?;
+
+        Ok(response)
+    }
+
+    fn start_iterable_query(
```

Comment:
`iterable` should be the default and the fact that it's query is implicit in the trait name

Reply:
Can agree on the iterable part, but removing query from the name is a bad idea. `Client` implements this trait, so, if one imports the query, this method can get exposed on it (even though it is not the main API to use). Because of that I made the name more specific.